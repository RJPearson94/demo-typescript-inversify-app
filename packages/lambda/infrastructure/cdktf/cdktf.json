{
  "language": "typescript",
  "app": "ts-node bin/app.ts",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
