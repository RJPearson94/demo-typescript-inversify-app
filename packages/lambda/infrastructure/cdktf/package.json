{
  "name": "@rjpearson94/lambda-cdktf-infrastructure",
  "description": "CDKTF infrastructure to deploy the Inversify function",
  "version": "1.0.0",
  "engines": {
    "node": ">= 14.15"
  },
  "scripts": {
    "get": "cdktf get",
    "deploy": "cdktf deploy",
    "destroy": "cdktf destroy",
    "synth": "cdktf synth",
    "test": "run-s test:unit",
    "test:unit": "jest ./test",
    "test:unit:snapshot-refresh": "jest ./test -u"
  },
  "devDependencies": {
    "@cdktf/provider-aws": "^1.0.70",
    "@types/jest": "^26.0.23",
    "@types/node": "^14.14.37",
    "cdktf": "0.3.0",
    "cdktf-cli": "0.3.0",
    "constructs": "^3.3.75",
    "jest": "^27.2.5",
    "npm-run-all": "^4.1.5",
    "ts-jest": "^26.5.6",
    "ts-node": "^9.1.1",
    "typescript": "~4.1.5"
  }
}
