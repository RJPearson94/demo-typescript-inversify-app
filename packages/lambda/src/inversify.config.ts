import { container } from '@rjpearson94/core';

import { TYPES } from '@src/constant/types';
import { GreetingController, GreetingControllerImpl } from '@src/controller';

container.bind<GreetingController>(TYPES.GreetingController).to(GreetingControllerImpl);

export { container };
