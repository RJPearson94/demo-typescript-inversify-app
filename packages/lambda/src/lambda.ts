import 'reflect-metadata';

import container from '@src/inversify.config';
import TYPES from '@src/constant/types';
import GreetingController from '@src/controller/greeting';

const greetingController: GreetingController = container.get<GreetingController>(TYPES.GreetingController);

exports.handler = function(event: any, context: any, callback: Function): void {
  const helloResponse = greetingController.greet();
  callback(null, {
    statusCode: 200,
    body: JSON.stringify({
      message: helloResponse
    })
  });
};
