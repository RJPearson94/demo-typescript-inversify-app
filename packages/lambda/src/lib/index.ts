export * from './apiResponse';
export * from './messageResponse';
export * from './context';