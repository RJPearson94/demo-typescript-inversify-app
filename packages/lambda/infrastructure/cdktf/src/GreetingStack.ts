import { Construct } from 'constructs';
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { TerraformStack, TerraformOutput } from 'cdktf';
import {
  LambdaFunction,
  CloudwatchLogGroup,
  IamRole,
  LambdaAlias,
  LambdaPermission,
  IamPolicy,
  IamRolePolicyAttachment,
  ApiGatewayRestApi,
  ApiGatewayDeployment,
  ApiGatewayStage,
  ApiGatewayApiKey,
  ApiGatewayUsagePlan,
  ApiGatewayUsagePlanKey
} from '@cdktf/provider-aws';
import openApiTemplate from './templates/openapi.tpl.json';

export type FunctionResources = {
  func: LambdaFunction;
  alias: LambdaAlias;
};

export interface ApiGatewayResources {
  restApi: ApiGatewayRestApi;
  stage: ApiGatewayStage;
  apiKey: ApiGatewayApiKey;
}

export type GreetingStackProps = {
  resourcePrefix?: string;
  api?: {
    stageName?: string;
    quota?: {
      limit?: number;
      period?: string;
    };
  };
  lambda?: {
    alias?: string;
  };
  tags?: { [name: string]: string };
};

export class GreetingStack extends TerraformStack {
  public readonly apiGatewayURL: TerraformOutput;
  public readonly apiKeyArn: TerraformOutput;

  private resourcePrefix: string;

  constructor(scope: Construct, private name: string, private props?: GreetingStackProps) {
    super(scope, name);

    this.resourcePrefix = props?.resourcePrefix ? `${props?.resourcePrefix}-` : '';

    const { alias } = this.createFunction();
    const { stage, apiKey } = this.createAPIGateway(alias);

    new LambdaPermission(this, `api_gateway_invoke_lambda`, {
      statementId: 'AllowAPIGatewayInvoke',
      action: 'lambda:InvokeFunction',
      functionName: alias.arn,
      principal: 'apigateway.amazonaws.com',
      sourceArn: `${stage.executionArn}/*`
    });

    this.apiGatewayURL = new TerraformOutput(this, `api_gateway_url`, {
      value: stage.invokeUrl
    });
    this.apiKeyArn = new TerraformOutput(this, `api_key_arn`, {
      value: apiKey.arn
    });
  }

  private createFunction(): FunctionResources {
    const functionName = `${this.resourcePrefix}${this.name}-function`;

    const logGroup = new CloudwatchLogGroup(this, `lambda_logs`, {
      name: `/aws/lambda/${functionName}`,
      retentionInDays: 7,
      tags: this.props?.tags
    });

    const lambdaRole = new IamRole(this, `lambda_role`, {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Principal: {
              Service: 'lambda.amazonaws.com'
            },
            Effect: 'Allow'
          }
        ]
      }),
      name: `${functionName}-role`,
      tags: this.props?.tags
    });

    const lambdaPolicy = new IamPolicy(this, `lambda_policy`, {
      name: `${functionName}-policy`,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: ['logs:CreateLogStream', 'logs:PutLogEvents'],
            Resource: logGroup.arn,
            Effect: 'Allow'
          },
          {
            Action: ['xray:PutTraceSegments', 'xray:PutTelemetryRecords'],
            Resource: '*',
            Effect: 'Allow'
          }
        ]
      })
    });

    new IamRolePolicyAttachment(this, `lambda_policy_attachment`, {
      role: lambdaRole.name,
      policyArn: lambdaPolicy.arn
    });

    const lambdaArtefactPath = '../../../dist/lambda.zip';
    const fileContents = fs.readFileSync(path.resolve(__dirname, lambdaArtefactPath));

    const func = new LambdaFunction(this, `lambda`, {
      filename: lambdaArtefactPath,
      sourceCodeHash: crypto.createHash('sha256').update(fileContents).digest('base64'),

      functionName,
      role: lambdaRole.arn,

      runtime: 'nodejs14.x',
      handler: 'main.handler',
      memorySize: 128,
      timeout: 5,
      tags: this.props?.tags,

      tracingConfig: [
        {
          mode: 'PassThrough'
        }
      ]
    });

    const alias = new LambdaAlias(this, `lambda_alias`, {
      name: this.props?.lambda?.alias || 'stable',
      functionName: func.functionName,
      functionVersion: func.version
    });

    return {
      func,
      alias
    };
  }

  private createAPIGateway(alias: LambdaAlias): ApiGatewayResources {
    const template = JSON.stringify(openApiTemplate).replace('${lambda_excution_uri}', alias.invokeArn);

    const apiGatewayName = `${this.resourcePrefix}${this.name}`;

    const restApi = new ApiGatewayRestApi(this, `rest_api`, {
      name: `${apiGatewayName}-rest-api`,
      body: template,
      tags: this.props?.tags
    });

    const deployment = new ApiGatewayDeployment(this, `deployment`, {
      restApiId: restApi.id,

      triggers: {
        redeployment: crypto.createHash('sha512').update(template).digest('hex')
      },

      lifecycle: {
        createBeforeDestroy: true
      }
    });

    const stage = new ApiGatewayStage(this, `stage`, {
      stageName: this.props?.api?.stageName || 'test',
      restApiId: restApi.id,
      deploymentId: deployment.id,
      xrayTracingEnabled: true,
      tags: this.props?.tags
    });

    const apiKey = new ApiGatewayApiKey(this, `api_key`, {
      name: `${apiGatewayName}-api-key`,
      tags: this.props?.tags
    });

    const usagePlan = new ApiGatewayUsagePlan(this, `usage_plan`, {
      name: `${apiGatewayName}-usage-plan`,
      apiStages: [
        {
          apiId: restApi.id,
          stage: stage.stageName
        }
      ],
      quotaSettings: [
        {
          limit: 100,
          period: 'MONTH',
          ...(this.props?.api?.quota || {})
        }
      ],
      tags: this.props?.tags
    });

    new ApiGatewayUsagePlanKey(this, `usage_plan_key`, {
      keyId: apiKey.id,
      usagePlanId: usagePlan.id,
      keyType: 'API_KEY'
    });

    return {
      restApi,
      stage,
      apiKey
    };
  }
}
