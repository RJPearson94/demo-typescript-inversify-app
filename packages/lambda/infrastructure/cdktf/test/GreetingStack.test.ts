import { Testing } from 'cdktf';
import { GreetingStack } from '../src/GreetingStack';

test('create greeting stack', () => {
  const app = Testing.app();
  const stack = new GreetingStack(app, 'greeting-stack');
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('create greeting stack with prefix and tags', () => {
  const app = Testing.app();
  const stack = new GreetingStack(app, 'greeting-stack', {
    resourcePrefix: 'test',
    tags: {
      service: 'greeting-stack',
      testing: 'true'
    }
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});
