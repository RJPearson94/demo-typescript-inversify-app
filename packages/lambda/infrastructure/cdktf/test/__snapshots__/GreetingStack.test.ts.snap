// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create greeting stack 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"greeting-stack\\"
    }
  },
  \\"resource\\": {
    \\"aws_cloudwatch_log_group\\": {
      \\"lambda_logs\\": {
        \\"name\\": \\"/aws/lambda/greeting-stack-function\\",
        \\"retention_in_days\\": 7,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_logs\\",
            \\"uniqueId\\": \\"lambda_logs\\"
          }
        }
      }
    },
    \\"aws_iam_role\\": {
      \\"lambda_role\\": {
        \\"assume_role_policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":\\\\\\"sts:AssumeRole\\\\\\",\\\\\\"Principal\\\\\\":{\\\\\\"Service\\\\\\":\\\\\\"lambda.amazonaws.com\\\\\\"},\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\",
        \\"name\\": \\"greeting-stack-function-role\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_role\\",
            \\"uniqueId\\": \\"lambda_role\\"
          }
        }
      }
    },
    \\"aws_iam_policy\\": {
      \\"lambda_policy\\": {
        \\"name\\": \\"greeting-stack-function-policy\\",
        \\"policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":[\\\\\\"logs:CreateLogStream\\\\\\",\\\\\\"logs:PutLogEvents\\\\\\"],\\\\\\"Resource\\\\\\":\\\\\\"\${aws_cloudwatch_log_group.lambda_logs.arn}\\\\\\",\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"},{\\\\\\"Action\\\\\\":[\\\\\\"xray:PutTraceSegments\\\\\\",\\\\\\"xray:PutTelemetryRecords\\\\\\"],\\\\\\"Resource\\\\\\":\\\\\\"*\\\\\\",\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_policy\\",
            \\"uniqueId\\": \\"lambda_policy\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"lambda_policy_attachment\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.lambda_policy.arn}\\",
        \\"role\\": \\"\${aws_iam_role.lambda_role.name}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_policy_attachment\\",
            \\"uniqueId\\": \\"lambda_policy_attachment\\"
          }
        }
      }
    },
    \\"aws_lambda_function\\": {
      \\"lambda\\": {
        \\"filename\\": \\"../../../dist/lambda.zip\\",
        \\"function_name\\": \\"greeting-stack-function\\",
        \\"handler\\": \\"main.handler\\",
        \\"memory_size\\": 128,
        \\"role\\": \\"\${aws_iam_role.lambda_role.arn}\\",
        \\"runtime\\": \\"nodejs14.x\\",
        \\"source_code_hash\\": \\"filebase64sha256(../../../dist/lambda.zip)\\",
        \\"timeout\\": 5,
        \\"tracing_config\\": [
          {
            \\"mode\\": \\"PassThrough\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda\\",
            \\"uniqueId\\": \\"lambda\\"
          }
        }
      }
    },
    \\"aws_lambda_alias\\": {
      \\"lambda_alias\\": {
        \\"function_name\\": \\"\${aws_lambda_function.lambda.function_name}\\",
        \\"function_version\\": \\"\${aws_lambda_function.lambda.version}\\",
        \\"name\\": \\"stable\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_alias\\",
            \\"uniqueId\\": \\"lambda_alias\\"
          }
        }
      }
    },
    \\"aws_api_gateway_rest_api\\": {
      \\"rest_api\\": {
        \\"body\\": \\"{\\\\\\"openapi\\\\\\":\\\\\\"3.0.1\\\\\\",\\\\\\"info\\\\\\":{\\\\\\"title\\\\\\":\\\\\\"Greeting API\\\\\\",\\\\\\"version\\\\\\":\\\\\\"1.0.0\\\\\\"},\\\\\\"paths\\\\\\":{\\\\\\"/v1/greet\\\\\\":{\\\\\\"get\\\\\\":{\\\\\\"summary\\\\\\":\\\\\\"Provide Greeting Message\\\\\\",\\\\\\"description\\\\\\":\\\\\\"Provide Greeting Message when lambda is invoked. Lambda is written in TypeScript\\\\\\",\\\\\\"operationId\\\\\\":\\\\\\"getV1Greet\\\\\\",\\\\\\"security\\\\\\":[{\\\\\\"ApiKey\\\\\\":[]}],\\\\\\"x-amazon-apigateway-integration\\\\\\":{\\\\\\"uri\\\\\\":\\\\\\"\${aws_lambda_alias.lambda_alias.invoke_arn}\\\\\\",\\\\\\"timeoutInMillis\\\\\\":10000,\\\\\\"httpMethod\\\\\\":\\\\\\"POST\\\\\\",\\\\\\"type\\\\\\":\\\\\\"aws_proxy\\\\\\",\\\\\\"responses\\\\\\":{\\\\\\"default\\\\\\":{\\\\\\"statusCode\\\\\\":\\\\\\"200\\\\\\"}}},\\\\\\"responses\\\\\\":{\\\\\\"200\\\\\\":{\\\\\\"description\\\\\\":\\\\\\"Successfully Greet\\\\\\",\\\\\\"content\\\\\\":{\\\\\\"application/json\\\\\\":{\\\\\\"schema\\\\\\":{\\\\\\"$ref\\\\\\":\\\\\\"#/components/schemas/GreetingResponse\\\\\\"}}}}}}}},\\\\\\"components\\\\\\":{\\\\\\"securitySchemes\\\\\\":{\\\\\\"ApiKey\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"apiKey\\\\\\",\\\\\\"in\\\\\\":\\\\\\"header\\\\\\",\\\\\\"name\\\\\\":\\\\\\"x-api-key\\\\\\"}},\\\\\\"schemas\\\\\\":{\\\\\\"GreetingResponse\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"object\\\\\\",\\\\\\"properties\\\\\\":{\\\\\\"message\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}}}}}}\\",
        \\"name\\": \\"greeting-stack-rest-api\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/rest_api\\",
            \\"uniqueId\\": \\"rest_api\\"
          }
        }
      }
    },
    \\"aws_api_gateway_deployment\\": {
      \\"deployment\\": {
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
        \\"triggers\\": {
          \\"redeployment\\": \\"12bc2a45ff45af9e0e33552ea15e29134514dd3b6f3c921dfa2a5c44a240bb8683d6b6468593b06c2248407901551c606d876c71da8a64729699b0fe531edd2d\\"
        },
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/deployment\\",
            \\"uniqueId\\": \\"deployment\\"
          }
        }
      }
    },
    \\"aws_api_gateway_stage\\": {
      \\"stage\\": {
        \\"deployment_id\\": \\"\${aws_api_gateway_deployment.deployment.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
        \\"stage_name\\": \\"test\\",
        \\"xray_tracing_enabled\\": true,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/stage\\",
            \\"uniqueId\\": \\"stage\\"
          }
        }
      }
    },
    \\"aws_api_gateway_api_key\\": {
      \\"api_key\\": {
        \\"name\\": \\"greeting-stack-api-key\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/api_key\\",
            \\"uniqueId\\": \\"api_key\\"
          }
        }
      }
    },
    \\"aws_api_gateway_usage_plan\\": {
      \\"usage_plan\\": {
        \\"name\\": \\"greeting-stack-usage-plan\\",
        \\"api_stages\\": [
          {
            \\"api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
            \\"stage\\": \\"\${aws_api_gateway_stage.stage.stage_name}\\"
          }
        ],
        \\"quota_settings\\": [
          {
            \\"limit\\": 100,
            \\"period\\": \\"MONTH\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/usage_plan\\",
            \\"uniqueId\\": \\"usage_plan\\"
          }
        }
      }
    },
    \\"aws_api_gateway_usage_plan_key\\": {
      \\"usage_plan_key\\": {
        \\"key_id\\": \\"\${aws_api_gateway_api_key.api_key.id}\\",
        \\"key_type\\": \\"API_KEY\\",
        \\"usage_plan_id\\": \\"\${aws_api_gateway_usage_plan.usage_plan.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/usage_plan_key\\",
            \\"uniqueId\\": \\"usage_plan_key\\"
          }
        }
      }
    },
    \\"aws_lambda_permission\\": {
      \\"api_gateway_invoke_lambda\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"function_name\\": \\"\${aws_lambda_alias.lambda_alias.arn}\\",
        \\"principal\\": \\"apigateway.amazonaws.com\\",
        \\"source_arn\\": \\"\${aws_api_gateway_stage.stage.execution_arn}/*\\",
        \\"statement_id\\": \\"AllowAPIGatewayInvoke\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/api_gateway_invoke_lambda\\",
            \\"uniqueId\\": \\"api_gateway_invoke_lambda\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"api_gateway_url\\": {
      \\"value\\": \\"\${aws_api_gateway_stage.stage.invoke_url}\\"
    },
    \\"api_key_arn\\": {
      \\"value\\": \\"\${aws_api_gateway_api_key.api_key.arn}\\"
    }
  }
}"
`;

exports[`create greeting stack with prefix and tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"greeting-stack\\"
    }
  },
  \\"resource\\": {
    \\"aws_cloudwatch_log_group\\": {
      \\"lambda_logs\\": {
        \\"name\\": \\"/aws/lambda/test-greeting-stack-function\\",
        \\"retention_in_days\\": 7,
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_logs\\",
            \\"uniqueId\\": \\"lambda_logs\\"
          }
        }
      }
    },
    \\"aws_iam_role\\": {
      \\"lambda_role\\": {
        \\"assume_role_policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":\\\\\\"sts:AssumeRole\\\\\\",\\\\\\"Principal\\\\\\":{\\\\\\"Service\\\\\\":\\\\\\"lambda.amazonaws.com\\\\\\"},\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\",
        \\"name\\": \\"test-greeting-stack-function-role\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_role\\",
            \\"uniqueId\\": \\"lambda_role\\"
          }
        }
      }
    },
    \\"aws_iam_policy\\": {
      \\"lambda_policy\\": {
        \\"name\\": \\"test-greeting-stack-function-policy\\",
        \\"policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":[\\\\\\"logs:CreateLogStream\\\\\\",\\\\\\"logs:PutLogEvents\\\\\\"],\\\\\\"Resource\\\\\\":\\\\\\"\${aws_cloudwatch_log_group.lambda_logs.arn}\\\\\\",\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"},{\\\\\\"Action\\\\\\":[\\\\\\"xray:PutTraceSegments\\\\\\",\\\\\\"xray:PutTelemetryRecords\\\\\\"],\\\\\\"Resource\\\\\\":\\\\\\"*\\\\\\",\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_policy\\",
            \\"uniqueId\\": \\"lambda_policy\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"lambda_policy_attachment\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.lambda_policy.arn}\\",
        \\"role\\": \\"\${aws_iam_role.lambda_role.name}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_policy_attachment\\",
            \\"uniqueId\\": \\"lambda_policy_attachment\\"
          }
        }
      }
    },
    \\"aws_lambda_function\\": {
      \\"lambda\\": {
        \\"filename\\": \\"../../../dist/lambda.zip\\",
        \\"function_name\\": \\"test-greeting-stack-function\\",
        \\"handler\\": \\"main.handler\\",
        \\"memory_size\\": 128,
        \\"role\\": \\"\${aws_iam_role.lambda_role.arn}\\",
        \\"runtime\\": \\"nodejs14.x\\",
        \\"source_code_hash\\": \\"filebase64sha256(../../../dist/lambda.zip)\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"timeout\\": 5,
        \\"tracing_config\\": [
          {
            \\"mode\\": \\"PassThrough\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda\\",
            \\"uniqueId\\": \\"lambda\\"
          }
        }
      }
    },
    \\"aws_lambda_alias\\": {
      \\"lambda_alias\\": {
        \\"function_name\\": \\"\${aws_lambda_function.lambda.function_name}\\",
        \\"function_version\\": \\"\${aws_lambda_function.lambda.version}\\",
        \\"name\\": \\"stable\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/lambda_alias\\",
            \\"uniqueId\\": \\"lambda_alias\\"
          }
        }
      }
    },
    \\"aws_api_gateway_rest_api\\": {
      \\"rest_api\\": {
        \\"body\\": \\"{\\\\\\"openapi\\\\\\":\\\\\\"3.0.1\\\\\\",\\\\\\"info\\\\\\":{\\\\\\"title\\\\\\":\\\\\\"Greeting API\\\\\\",\\\\\\"version\\\\\\":\\\\\\"1.0.0\\\\\\"},\\\\\\"paths\\\\\\":{\\\\\\"/v1/greet\\\\\\":{\\\\\\"get\\\\\\":{\\\\\\"summary\\\\\\":\\\\\\"Provide Greeting Message\\\\\\",\\\\\\"description\\\\\\":\\\\\\"Provide Greeting Message when lambda is invoked. Lambda is written in TypeScript\\\\\\",\\\\\\"operationId\\\\\\":\\\\\\"getV1Greet\\\\\\",\\\\\\"security\\\\\\":[{\\\\\\"ApiKey\\\\\\":[]}],\\\\\\"x-amazon-apigateway-integration\\\\\\":{\\\\\\"uri\\\\\\":\\\\\\"\${aws_lambda_alias.lambda_alias.invoke_arn}\\\\\\",\\\\\\"timeoutInMillis\\\\\\":10000,\\\\\\"httpMethod\\\\\\":\\\\\\"POST\\\\\\",\\\\\\"type\\\\\\":\\\\\\"aws_proxy\\\\\\",\\\\\\"responses\\\\\\":{\\\\\\"default\\\\\\":{\\\\\\"statusCode\\\\\\":\\\\\\"200\\\\\\"}}},\\\\\\"responses\\\\\\":{\\\\\\"200\\\\\\":{\\\\\\"description\\\\\\":\\\\\\"Successfully Greet\\\\\\",\\\\\\"content\\\\\\":{\\\\\\"application/json\\\\\\":{\\\\\\"schema\\\\\\":{\\\\\\"$ref\\\\\\":\\\\\\"#/components/schemas/GreetingResponse\\\\\\"}}}}}}}},\\\\\\"components\\\\\\":{\\\\\\"securitySchemes\\\\\\":{\\\\\\"ApiKey\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"apiKey\\\\\\",\\\\\\"in\\\\\\":\\\\\\"header\\\\\\",\\\\\\"name\\\\\\":\\\\\\"x-api-key\\\\\\"}},\\\\\\"schemas\\\\\\":{\\\\\\"GreetingResponse\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"object\\\\\\",\\\\\\"properties\\\\\\":{\\\\\\"message\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}}}}}}\\",
        \\"name\\": \\"test-greeting-stack-rest-api\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/rest_api\\",
            \\"uniqueId\\": \\"rest_api\\"
          }
        }
      }
    },
    \\"aws_api_gateway_deployment\\": {
      \\"deployment\\": {
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
        \\"triggers\\": {
          \\"redeployment\\": \\"12bc2a45ff45af9e0e33552ea15e29134514dd3b6f3c921dfa2a5c44a240bb8683d6b6468593b06c2248407901551c606d876c71da8a64729699b0fe531edd2d\\"
        },
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/deployment\\",
            \\"uniqueId\\": \\"deployment\\"
          }
        }
      }
    },
    \\"aws_api_gateway_stage\\": {
      \\"stage\\": {
        \\"deployment_id\\": \\"\${aws_api_gateway_deployment.deployment.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
        \\"stage_name\\": \\"test\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"xray_tracing_enabled\\": true,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/stage\\",
            \\"uniqueId\\": \\"stage\\"
          }
        }
      }
    },
    \\"aws_api_gateway_api_key\\": {
      \\"api_key\\": {
        \\"name\\": \\"test-greeting-stack-api-key\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/api_key\\",
            \\"uniqueId\\": \\"api_key\\"
          }
        }
      }
    },
    \\"aws_api_gateway_usage_plan\\": {
      \\"usage_plan\\": {
        \\"name\\": \\"test-greeting-stack-usage-plan\\",
        \\"tags\\": {
          \\"service\\": \\"greeting-stack\\",
          \\"testing\\": \\"true\\"
        },
        \\"api_stages\\": [
          {
            \\"api_id\\": \\"\${aws_api_gateway_rest_api.rest_api.id}\\",
            \\"stage\\": \\"\${aws_api_gateway_stage.stage.stage_name}\\"
          }
        ],
        \\"quota_settings\\": [
          {
            \\"limit\\": 100,
            \\"period\\": \\"MONTH\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/usage_plan\\",
            \\"uniqueId\\": \\"usage_plan\\"
          }
        }
      }
    },
    \\"aws_api_gateway_usage_plan_key\\": {
      \\"usage_plan_key\\": {
        \\"key_id\\": \\"\${aws_api_gateway_api_key.api_key.id}\\",
        \\"key_type\\": \\"API_KEY\\",
        \\"usage_plan_id\\": \\"\${aws_api_gateway_usage_plan.usage_plan.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/usage_plan_key\\",
            \\"uniqueId\\": \\"usage_plan_key\\"
          }
        }
      }
    },
    \\"aws_lambda_permission\\": {
      \\"api_gateway_invoke_lambda\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"function_name\\": \\"\${aws_lambda_alias.lambda_alias.arn}\\",
        \\"principal\\": \\"apigateway.amazonaws.com\\",
        \\"source_arn\\": \\"\${aws_api_gateway_stage.stage.execution_arn}/*\\",
        \\"statement_id\\": \\"AllowAPIGatewayInvoke\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/api_gateway_invoke_lambda\\",
            \\"uniqueId\\": \\"api_gateway_invoke_lambda\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"api_gateway_url\\": {
      \\"value\\": \\"\${aws_api_gateway_stage.stage.invoke_url}\\"
    },
    \\"api_key_arn\\": {
      \\"value\\": \\"\${aws_api_gateway_api_key.api_key.arn}\\"
    }
  }
}"
`;
